### DISCUSSION TITLE
[Feature] GPX support

### DISCUSSION THREAD

**Discussion by Terramoto** [immich-app/immich#6831] [Feature Request]
### The feature

Although it might be off-topic, but considering the app has a map and makes use of photo geo-location, i thought it would be a nice addition to be able to upload/import/use gpx tracks like the ones generated from Osmand trip recording, you can already take pictures and generate waypoints with the <name>pictureName.jpg</name> tag in the gpx file.

i'm currently using a modified wtracks (gpx viewer) to host my mushroom hunting trips but i would love to use a more picture focused tool to manage the pictures along with the tracks.

![Imagem WhatsApp 2024-02-01 às 09 17 15_f8c32023](https://github.com/immich-app/immich/assets/5443354/31c33d98-0736-4018-a587-49913864b7bd)

### Platform

- [X] Server
- [X] Web
- [ ] Mobile

**Comment by AngelaDMerkel** [comment]
I think this would be a nice feature addition, but only if Immich works to support additional location related features like interpolation between points etc. If you want to match precise data to your images it is possible to do so by writing GPS data into the file using exiftool or another similar utility.

**Comment by Terramoto** [comment]
> I think this would be a nice feature addition, but only if Immich works to support additional location related features like interpolation between points etc. If you want to match precise data to your images it is possible to do so by writing GPS data into the file using exiftool or another similar utility.
Are you referring to something like this?
https://maplibre.org/maplibre-gl-js/docs/examples/data-driven-lines/

Images in the map could be filtered on clicking the track (by date and gps location) or vice versa.

**Reply by AngelaDMerkel** [reply]
> Are you referring to something like this? https://maplibre.org/maplibre-gl-js/docs/examples/data-driven-lines/

No, I am not. GPX files usually only collect data at given intervals, assuming they have location information to collect at all. As a result, the GPS data very rarely matches up with images and it's necessary to guess where the images would have been taken based on factors including a middle point between the a known GPS location before and after the image was taken. This requires some interpolation or some other mechanism for assessing the likelihood of an image having been taken at a given location.

**Comment by gardiol** [comment]
I used to use this kind of feature on Digikam, which is a great feature and very useful for people who travels a lot and takes tracks of the trip. I literally have tens of thousands of pictures geolocalized in this way (how Digikam did the matching i have no idea, but it worked quite well) and it would be an awesome feature for Immich.

Even using "middle points" or interpolation would be great as precision is not really needed, it's more important to pinpoint each photo (of lots) to the correct more-or-less area. Consider that 99% of the time when you take pictures you also stop, so there is usually a few points in the track very close to each bunch of pictures. The remaining ones, we can live with some manual adjustment.

**Reply by Terramoto** [reply]
OSMand seems to attach a <name> tag with the picture filename in the gpx file. It could also match the date and time in the exif with the time in the GPX file down to the second which would be more reliable than interpolating GPS data. At least the experience i have with my phone is that it doesn't wait for a GPS lock, it just pings and eventually gets a cached data (miles away) but it might be specific to my device.

**Reply by gardiol** [reply]
Yes, phone provided location in photos is often inaccurate i agree, that's why in my travels i never care, but add my GPX tracks afterward. 

Also, DSLR cameras often just don't add any GPS data, but this is for "external" libraries in Immich anyway and will be managed externally from Immitch itself.

Never used OSMand, i use AlpineQuest or the Garmin Explore apps, they both generate nice tracks which can be used easily to geotag photos.

**Reply by AngelaDMerkel** [reply]
The main issue is that this feature must be implemented with some kind of interpolation, otherwise photo time stamps with virtually never line up with the GPS polling times. This is especially true which activities like hiking where GPS can fall in and out constantly and make GPS wildly inaccurate.

**Reply by gardiol** [reply]
As i said, Digikam does that and works very good. I have been geotagging photos for the last 10 years with GPX tracks recorded with the most various devices (including, back in time, old Nokia 5800 XM) from very detailed to very sparse tracks. I never had an issue with interpolation. Yes, sometimes you need to adjust some photos, specially when taken inside, but nobody expect a flawless and perfect solution.  So, from my experience, interpolation works and does not cause issues.

**Reply by AngelaDMerkel** [reply]
@gardiol Digikam is a useful tool, but it doesn't fit well into an existing workflow, especially one which uses RAW or one which draws from a wide variety of sources. If multiple users are using Immich, especially less tech savvy ones, you can't expect them to download this application to tag their photos before they share them with you, and it is equally as frustrating to be required to download their images shared on Immich only to reupload. 

I rely on exiftool to handle tagging from my gpx files and don't mind the additional nuisance, but no one in my family is going to bother with that.

**Reply by gardiol** [reply]
I don't mean to use digikam of course... I was simply stating that gpx import with interpolation not only works but would be a great addition to Immich.

**Reply by aisbergde** [reply]
And I have am opposite opinion: Of course this could be implemented into immich, but this also means, it needs to be improved and maintained. On the other side, there are external solutions, specialized on geo tag management, and they are doing well.

That's why I think it is much more important, to improve the way, how immich can interact witch external asset management (recognition of external deletions and changes in the default libraries, allowing to delete assets in external libraries via GUI).

**Reply by gardiol** [reply]
I fully agree its low priority, there are other more important things to address before this one! Nevertheless, when immich will be so refined that this feature will be worked on, I think will add value to immich.

**Comment by bonerlog** [comment]
Adding a GPX or FIT file (most sport trackers use FIT) to an album and then showing the track and thumbnails of the photos on a map (position based on timestamps/EXIF) would be great. For external libraries this could be done by just scanning if folders include a GPX/FIT.

When a photo that belongs to a track (part of album/folder with GPX file and timestamp matches) gets maximized in the timeline, the track should appear, too.

If this gets implemented it would be important to have the ability to set time offsets. I have attached a screenshot of JOSM for example.

Geotagging as an option would also be nice to have. But there are already external solutions for that, so that's not needed necessarily I think.

![GPX-support](https://github.com/immich-app/immich/assets/158835976/98e1f578-01fa-4885-b428-32f65c9846ca)

**Reply by aisbergde** [reply]
showing the track and thumbnails of the photos on a map - good idea. but without using any additional file, and only based on the information stored in assets.

**Reply by bonerlog** [reply]
I think what you want needs a new topic as it has nothing to do with GPX support. The idea in this thread is based on an additional file, which means the track has already been recorded very precisely using GPS. Matching works by just using the timestamps, so there is no need for a geotag in the EXIF data itself.

What you want would look more like this, using already geotagged images without a track file:
![geotagged](https://github.com/immich-app/immich/assets/158835976/64e196e7-1062-42dc-94e6-883d958c3cf2)

**Reply by aisbergde** [reply]
Exactly, it is just an additional idea, based on this thread here. But of course, I get the difference: a more continuous presentation vs a discrete presentation only for existing assets. They could even be combined.
The idea was born, because it could be implemented using only the already existing data and existing map implementation.

**Comment by bo0tzz** [comment]
https://dawarich.app/ is built for this and has an Immich integration.

**Reply by AngelaDMerkel** [reply]
This is spectacular. Thanks for drawing our attention to it

**Comment by frostworx** [comment]
yes, dawarich rocks, I also suggest https://github.com/Flomp/wanderer (which can automatically import komoot gpx files since a recent update).

for my dslr cameras without gps, I currently use https://github.com/FrancoisSchnell/GPicSync to add the coordinates from the gpx file to the pictures in a temp dir. Afterward, I copy the pictures _(now with the coordinates in exif)_ to the immich external library.

A few extra steps, but this works just fine.

**Reply by AngelaDMerkel** [reply]
I use GPX logs and an application called HoudahGeo to write the gps to exif before importing into LrC. It's good because it also has integration with commercial maps to tag where the GPX is data is missing or inaccurate as is prone to happen. It's a very good tool.

**Reply by frostworx** [reply]
Thanks for mentioning the tool! I'm on Linux, so unfortunately no option for me, but I'll tell my wife, she dualboots ;)
Nice nick btw,  @AngelaDMerkel :)

**Comment by frostworx** [comment]
sry for the necrobump. I recently found this tool which directly reads the geolocation from dawarich and applies it to your matching pictures: https://github.com/streiman/lazygeotag

edit: just found out that dawarich already has the "Enrich Photos" option, which deprecates the need for lazygeotag

**Comment by Griffsano** [comment]
Hi everyone,

I wanted to share a related tool that I recently implemented: https://github.com/Griffsano/immich2gpx

It does not add native GPX support to Immich, but it may help with some of the workflows discussed here. **immich2gpx** queries media metadata from an Immich server and exports the EXIF location data of selected photos as GPX tracks.

I built it mainly because the existing Immich import feature for Dawarich did not fit my use case: my Immich library contains shared geotagged photos, including photos taken by other people at locations where I was not present, which can affect my personal location history.

With **immich2gpx**, you can use Immich filters to export only the photos relevant to a specific activity or journey and then use the generated GPX file with tools like Dawarich or other GPX-based tools.

Feel free to give it a try if it fits your use case. Feedback and suggestions are very welcome.

Thanks for the great work on Immich!
